### ISSUE TITLE
What are the determinants of comm > nchannels?

### ISSUE THREAD

**Issue by xiabo123** [NVIDIA/nccl#524]
In the function ncclTopoPostset, I found that comm > nchannels has been set many times. What factors are the factors that influence comm > nchannels?

Thank you very much!

**Comment by sjeaugey** [comment]
The determination of `comm->nchannels` is indeed a complex thing. First we set it to the number of channels the graph search discovered based on the node topology. Then we double the number of channels (during ncclTopoPostset) to increase efficiency.
And finally we honor what the user may have set as min/max nchannels.

**Comment by xiabo123** [comment]
Thanks for your response!

**Comment by xiabo123** [comment]
Thanks for your response!

**Comment by xiabo123** [comment]
@sjeaugey Hello, ask again, is this doubling fixed? Is quadruple OK?

**Comment by sjeaugey** [comment]
Yes it is a fixed number. You could quadruple, but that would use twice a many GPU SMs when running.

**Comment by xiabo123** [comment]
Thanks for your response!
yes，I modified it, but it didn't improve the performance.
I have another question. When communicating within a node, can the sequence of ring be modified? Or can I add a custom ring sequence?
Thank you very much!

**Comment by sjeaugey** [comment]
Yes, you can dump the rings and trees created with `NCCL_GRAPH_DUMP_FILE=graph.xml`, then modify the generated file, and re-inject it with `NCCL_GRAPH_FILE=graph.xml`.

**Comment by xiabo123** [comment]
Thanks for your response!

**Comment by xiabo123** [comment]
@sjeaugey Hello, ask again, Can the Infiniband between nodes be specified manually?

**Comment by sjeaugey** [comment]
I'm not sure what part of "Infiniband" you want to specify manually.

In the `graph.xml`, you can specify which network interface to use for each ring. `<graph id=0>` describes the intra-node order for the rings. When running inter-node, each `<channel>` should start and end with `<net dev=X>`. You can specify the `X` to the index of the Infiniband interface you want to use. Similarly, `<graph id=1>` describes the intra-node sequence for the trees.

You cannot define the order of the nodes in the ring/tree however.

Hope this answers your question.

**Comment by xiabo123** [comment]
Sorry, ask again.
![image](https://user-images.githubusercontent.com/40659418/126321879-5c167c01-376b-4675-bf3f-e8d53985557d.png)
What does "the order of the nodes in the ring/tree" mean? 
I understand that it refers to the "      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>" section in the above figure, right?

**Comment by sjeaugey** [comment]
The order of the ring is :

1. `<net dev="3">` Data comes from the previous node in the ring through network interface 3 (see the topology with `NCCL_TOPO_DUMP_FILE=system.xml` to see which interface has an `id="3"`)
2. `<gpu dev="0">`, `<gpu dev="1"> ` `<gpu dev="2">`, `<gpu dev="3">`  data goes from network 3 to GPU 0, then GPU 1, GPU 2 and GPU 3
3. `<net dev="0">` then data goes from GPU 3 to network interface 0 onto the next node. Given `crossnic=0`, I assume network interfaces 0 and 3 are 2 interfaces of the same NIC which has 2 ports and is connected to the node through 2 different PCI connections.

I would also assume that network interface 3 is close to GPU 0 and network interface 0 is close to GPU 3, which you can confirm looking at the system topology.

**Comment by xiabo123** [comment]
Thanks for your response!

**Comment by xiabo123** [comment]
I have another doubt. As shown in the figure above, there is only one ring.Data goes from network 3 to GPU 0, then GPU 1, GPU 2 and GPU 3. Network interface 3 is close to GPU 0 and network interface 0 is close to GPU 3.Network interface 2 is close to GPU 1 and network interface 1 is close to GPU 2.
If I want to use network interface 1 and network interface 2, what should I do.

**Comment by xiabo123** [comment]
In the node, ring can be added ,use twice a many GPU SMs when running .Can I use different network interface ports when switching between nodes?

**Comment by sjeaugey** [comment]
You could add another `<channel>` and update `nchannels=2`. Make sure you also add a channel and set `nchannels=2` in the trees (`graph id="1"`) section otherwise NCCL will use the min of rings and trees.

Now, keep in mind that if NCCL did not use all ports, there is likely a bottleneck somewhere in the topology and adding more rings may not improve performance significantly (or could even degrade performance, as well as GPU usage).

**Comment by xiabo123** [comment]
First of all, thank you for your reply.
how to modify graph.xml? I tried to modify it, but the result was not ideal.
The content of the XML file is as follows:
```
<graphs version="1">
  <graph id="0" pattern="4" crossnic="0" nchannels="1" speedintra="12" speedinter="12" typeintra="SYS" typeinter="SYS" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="0"/>
    </channel>
  </graph>
  <graph id="1" pattern="3" crossnic="0" nchannels="1" speedintra="6" speedinter="6" typeintra="SYS" typeinter="SYS" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="3"/>
    </channel>
  </graph>
  <graph id="2" pattern="3" crossnic="0" nchannels="0" speedintra="0" speedinter="0" typeintra="NVL" typeinter="PIX" samechannels="1"/>
</graphs>

```

**Comment by sjeaugey** [comment]
See an example below with 2 channels. You can change the order of the GPUs, and the network interfaces for data to enter/exit the node.
```
<graphs version="1">
  <graph id="0" pattern="4" crossnic="0" nchannels="2" speedintra="12" speedinter="12" typeintra="SYS" typeinter="SYS" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="0"/>
    </channel>
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="0"/>
    </channel>
  </graph>
  <graph id="1" pattern="3" crossnic="0" nchannels="2" speedintra="6" speedinter="6" typeintra="SYS" typeinter="SYS" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="3"/>
    </channel>
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="3"/>
    </channel>
  </graph>
  <graph id="2" pattern="3" crossnic="0" nchannels="0" speedintra="0" speedinter="0" typeintra="NVL" typeinter="PIX" samechannels="1"/>
</graphs>
```

**Comment by xiabo123** [comment]
yes.
I modified it like this, but the number of channels is still 1 and the number after double is 2.In theory, the number of channels should be 2, and double it should be 4.When in a node, this modification can increase the number of channels.
According to the log, the added network interface is not used.
The contents are as follows:
xml:
```
<graphs version="1">
  <graph id="0" pattern="4" crossnic="0" nchannels="2" speedintra="12" speedinter="12" typeintra="SYS" typeinter="SYS" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="0"/>
    </channel>
    <channel>
      <net dev="2"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="1"/>
    </channel>
  </graph>
  <graph id="1" pattern="3" crossnic="0" nchannels="2" speedintra="6" speedinter="6" typeintra="SYS" typeinter="SYS" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="3"/>
    </channel>
    <channel>
      <net dev="2"/>
      <gpu dev="0"/>
      <gpu dev="1"/>
      <gpu dev="2"/>
      <gpu dev="3"/>
      <net dev="2"/>
    </channel>
  </graph>
  <graph id="2" pattern="3" crossnic="0" nchannels="0" speedintra="0" speedinter="0" typeintra="NVL" typeinter="PIX" samechannels="1"/>
</graphs>
```
log
```
e05r3n07:14904:15137 [0] NCCL INFO Channel 00/02 :    0   1   2   3   4   5   6   7
e05r3n07:14904:15137 [0] NCCL INFO Channel 01/02 :    0   1   2   3   4   5   6   7
```

**Comment by sjeaugey** [comment]
That's weird indeed. Do you see an INFO line saying "Loading graphs"? Do you see any error in the log after that? 

Note (probably unrelated): in the XML, samechannels should probably be set to 0 given channels are not identical.

**Comment by xiabo123** [comment]
I didn't find the log you said.
Does this log matter?
`NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation`

**Comment by sjeaugey** [comment]
Did you set `NCCL_GRAPH_FILE=graph.xml`?

**Comment by xiabo123** [comment]
Sorry, this environment variable has been unset.
Now, Thanks for your response!

**Comment by xiabo123** [comment]
Another question: I should use network interface 3 and network interface 2 according to the following settings, but actually only network interface 2 has data transmission.
Network interface is used by the on the log.
```
<graphs version="1">
  <graph id="0" pattern="4" crossnic="0" nchannels="2" speedintra="12" speedinter="12" typeintra="LOC" typeinter="PHB" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <net dev="3"/>
    </channel>
    <channel>
      <net dev="2"/>
      <gpu dev="0"/>
      <net dev="2"/>
    </channel>
  </graph>
  <graph id="1" pattern="3" crossnic="0" nchannels="2" speedintra="24" speedinter="12" typeintra="LOC" typeinter="PHB" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <net dev="0"/>
    </channel>
    <channel>
      <net dev="2"/>
      <gpu dev="0"/>
      <net dev="2"/>
    </channel>
  </graph>
  <graph id="2" pattern="3" crossnic="0" nchannels="0" speedintra="0" speedinter="0" typeintra="LOC" typeinter="PIX" samechannels="1"/>
</graphs>
```
![image](https://user-images.githubusercontent.com/40659418/127295031-f7d282d3-5ee6-4543-b4bc-f8ef6adec047.png)

**Comment by sjeaugey** [comment]
I'm not sure why you would not see any traffic from one of the network interfaces given the logs does use both.

Is your network Infiniband or Ethernet(RoCE)? How do you monitor the data transmission?

**Comment by xiabo123** [comment]
I may not be clear. As mentioned above, I increased network interface 2, but the speed did not improve. I doubt that the increase 
 network interface 2 was not used.

**Comment by sjeaugey** [comment]
As I mentioned before,
> Now, keep in mind that if NCCL did not use all ports, there is likely a bottleneck somewhere in the topology and adding more rings may not improve performance significantly (or could even degrade performance, as well as GPU usage).

Can you post your node topology (set `NCCL_TOPO_DUMP_FILE=system.xml.txt` then post `system.xml.txt` here), or explain what performance you get by default and what you would expect to get using two NICs?

**Comment by xiabo123** [comment]
```
<graphs version="1">
  <graph id="0" pattern="4" crossnic="0" nchannels="1" speedintra="12" speedinter="12" typeintra="LOC" typeinter="PHB" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <net dev="3"/>
    </channel>
  </graph>
  <graph id="1" pattern="3" crossnic="0" nchannels="1" speedintra="24" speedinter="12" typeintra="LOC" typeinter="PHB" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <net dev="3"/>
    </channel>
  </graph>
  <graph id="2" pattern="3" crossnic="0" nchannels="0" speedintra="0" speedinter="0" typeintra="LOC" typeinter="PIX" samechannels="1"/>
</graphs>
```

**Comment by xiabo123** [comment]
Just like this, there is only one network interface. According to speedier = 12, the speed between nodes should be 12GB / s, and the test is only 5.5GB/s. I think increasing network interface should increase the bandwidth. But the effect is not ideal.
```
<graphs version="1">
  <graph id="0" pattern="4" crossnic="0" nchannels="1" speedintra="12" speedinter="12" typeintra="LOC" typeinter="PHB" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <net dev="3"/>
    </channel>
  </graph>
  <graph id="1" pattern="3" crossnic="0" nchannels="1" speedintra="24" speedinter="12" typeintra="LOC" typeinter="PHB" samechannels="1">
    <channel>
      <net dev="3"/>
      <gpu dev="0"/>
      <net dev="3"/>
    </channel>
  </graph>
  <graph id="2" pattern="3" crossnic="0" nchannels="0" speedintra="0" speedinter="0" typeintra="LOC" typeinter="PIX" samechannels="1"/>
</graphs>
```

**Comment by sjeaugey** [comment]
12 GB/s is the ideal performance; I would guess your NICs are 100Gbps. There could be many reasons why performance is not optimal and we're not even saturating one NIC. We should understand that first, and for that I'd need the system topology, or at least understand how NICs and GPUs are connected to each other (through the CPU, through a PCI switch, is there NVLink between the GPUs, is GPU Direct RDMA installed on the node if using PCI switches, ... all that information being inside the topology XML).

**Comment by xiabo123** [comment]
One more question:
The number of channels is calculated according to the topo structure. Is there any relevant information about the principle of this calculation?

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
